File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int solution(vector<vector<int>> jobs) {
    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>>pq;
    int answer = 0;
    int i = 0;
    int cnt = 0; //연속된 일의 시간을 확인
    
    sort(jobs.begin(),jobs.end());
    
    while(i < jobs.size() || !pq.empty()){
        if(i < jobs.size() && jobs[i][0] <= cnt){
            pq.push({jobs[i][1],jobs[i][0]});
            i++;
            continue;
        }
        if(!pq.empty()){
            cnt += pq.top().first;
            answer += cnt - pq.top().second;
            pq.pop();
        }
        else cnt = jobs[i][0];
    }
    answer /= jobs.size();
    
    return answer;
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    
    static class Pair{
        int x; int y; int z;
        public Pair(int x, int y, int z){
            //소요 시간, 요청 시각, 작업 번호
            this.x = x; this.y = y; this.z = z;
        }
    }
    
    public int solution(int[][] jobs) {
        int n = jobs.length;
        
        //입력: [작업이 요청되는 시점, 작업의 소요시간]
        Arrays.sort(jobs, (a,b) -> {
            if(a[0] == b[0]) return a[1] - b[1];
            else return a[0] - b[0];
        });
            
        PriorityQueue<Pair>pq = new PriorityQueue<>((a,b)->{
            if(a.x != b.x) return a.x - b.x;
            if(a.y != b.y) return a.y - b.y;
            return a.z - b.z;
        });
        
        int idx = 0, time = 0, answer = 0;
        while(idx < n || !pq.isEmpty()){
            while(idx < n && jobs[idx][0] <= time){
                pq.offer(new Pair(jobs[idx][1], jobs[idx][0], idx++));
            }
            
            if(pq.isEmpty()) {
                time = jobs[idx][0];
                continue;
            }
            
            Pair cur = pq.poll();
            time += cur.x;
            answer += time - cur.y;
        }
        
        return answer / n;
    }
}